buildSrc/src/main/groovy/org/elasticsearch/gradle/test/RestIntegTestTask.groovy

```diff
@@ -69,9 +69,21 @@ class RestIntegTestTask extends DefaultTask {
             runner.systemProperty('test.clustername', System.getProperty("tests.clustername"))
         }
 
-        // copy the rest spec/tests onto the test classpath
-        Copy copyRestSpec = createCopyRestSpecTask()
+        // copy the full rest spec (including modules, plugins, and x-pack) to a common location
```

Comment:
We probably don't want this logic to live in this task, for several reasons:

1. We shouldn't have _any_ logic in task constructors to begin with. We actually have an [existing issue](https://github.com/elastic/elasticsearch/issues/47804) to refactor this task.
1. We create _lots_ of these tasks, across many projects, so we'll end up creating a set of copy tasks for each project which defines a `RestIntegTestTask`. No bueno.
1. All these copy tasks share the same copy _destination_ which is going to cause all sorts of issues in parallel builds with concurrently executing copy tasks clobbering eachothers output.
1. Gradle will actually issue a warning because of issue (2).
1. We use `RestIntegTestTask` whenever we want to run integration tests against an external cluster. Firstly, not all these are YAML tests, and second, not all projects that use YAML tests also necessarily need to full rest test suite.

Here's what I would suggest we do instead.

1. Create a new plugin that we apply to the root project only to set all this stuff up. This ensure we only ever do this once. We could create this as a binary plugin but given the way we set this stuff up I'm inclined to start as a script plugin. That is. create a `rest-test-spec.gradle` file in the `gradle` folder that we then appy to the root build script via `appy from: 'gradle/rest-test-spect.gradle`.
2. Define the output of all these copy/unpack tasks as a build _artifact_. This will allow projects that need it to _depend_ on it, which is the correct way of modeling this, and also ensures stuff gets built only if necessary, and in the correct order. We can actually declare a _directory_ as an artifact (this is how compiled classes are depended on in fact), and you can see an example of that _here_.
3. We then in projects that require this stuff, declare a dependency on this stuff. Since we just need this stuff on the test runtime classpath, we'd so something like `testRuntime project(path: ':', configuration: 'test-rest-spec')` and then all this stuff will magically be copied and added to the target projects test runtime classpath. No more copying to output directories.

Please feel free to change any of these names to something that makes sense and reach out for questions.

Reply:
Thanks @mark-vieira, I exactly wasn't sure how to approach this (and still learning Gradle) so this helps alot.